src/relay/pass/pattern_util.h

```diff
@@ -204,7 +208,14 @@ template<typename T>
 inline Constant MakeConstantScalar(DataType dtype, T value) {
   runtime::NDArray arr = runtime::NDArray::Empty({}, Type2TVMType(dtype), {kDLCPU, 0});
   TVM_DTYPE_DISPATCH(dtype, DType, {
-    *static_cast<DType*>(arr->data) = value;
+    if (dtype == Float(16)) {
+      // convert to float16
+      // storage is uint16_t
+      *static_cast<DType*>(arr->data) =
+        __truncXfYf2__<float, uint32_t, 23, uint16_t, uint16_t, 10>(static_cast<float>(value));
```

Comment:
please add an assertion that `T` is `float32`

Reply:
@vinx13 
* static_cast<float>(T value) will make sure __truncXfyf2() is feeding with float32.
* unsure, still want to ```assert(T is_float32)``` or we let T to be any type ?

Reply:
This bit manipulation seems to bring a certain level of tediousness/ugliness in the TVM codebase.

A way to resolve this can be to keep the Float constants always in FP32. And insert a Relay cast operation to FP16 if need be. Then this will become part of fold_constant in Relay graph, and will use LLVM to generate the FP32 to FP16 conversion (hiding the bit manipulation).

I do not have any strong opinion on either. What do you think @vinx13 ?

Reply:
Because we support arbitrary bit types in the IR, we should really use gmp and mpfr for storing literals. It is bad practice to store them in finite types, as truncation may occur and it is impossible to test whether truncation occurred programmatically w/o weird hacks. This is out of the scope of this PR unfortunately.

Reply:
@cbalint13 ignore my comments :) the static cast if sufficient here